projector-agent/build.gradle.kts

```diff
@@ -0,0 +1,145 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact JetBrains, Na Hrebenech II 1718/10, Prague, 14000, Czech Republic
+ * if you need additional information or have any questions.
+ */
+
+import java.util.*
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+plugins {
+  kotlin("jvm")
+  `maven-publish`
+}
+
+publishing {
+  publications {
+    create<MavenPublication>("maven") {
+      from(components["java"])
+    }
+  }
+}
+
+kotlin {
+  explicitApi()
+}
+
+val compileKotlin: KotlinCompile by tasks
+val targetJvm: String by project
+compileKotlin.kotlinOptions.jvmTarget = targetJvm
+
+val projectorClientGroup: String by project
+val projectorClientVersion: String by project
+val javassistVersion: String by project
+val agentClassName = "org.jetbrains.projector.agent.MainAgent"
+val launcherClassName = "org.jetbrains.projector.agent.AgentLauncher"
+version = project(":projector-plugin").version
+
+dependencies {
+  implementation("$projectorClientGroup:projector-common:$projectorClientVersion")
+  implementation("$projectorClientGroup:projector-server-core:$projectorClientVersion")
+  implementation("$projectorClientGroup:projector-util-logging:$projectorClientVersion")
+  implementation("$projectorClientGroup:projector-util-loading:$projectorClientVersion")
+  api(project(":projector-awt"))
+  api(project(":projector-server"))
+  implementation("org.javassist:javassist:$javassistVersion")
+}
+
+var serverTargetClasspath: String? = null
+var serverClassToLaunch: String? = null
+var ideaPath: String? = null
+
+rootProject.file("local.properties").let {
+  if (it.canRead()) {
+    ideaPath = Properties().apply { load(it.inputStream()) }.getProperty("projectorLauncher.ideaPath") ?: null
+    serverTargetClasspath = Properties().apply { load(it.inputStream()) }.getProperty("projectorLauncher.targetClassPath") ?: null
+    serverClassToLaunch = Properties().apply { load(it.inputStream()) }.getProperty("projectorLauncher.classToLaunch") ?: null
+  }
+}
+
+tasks.withType<Jar> {
+  manifest {
+    attributes(
+      "Can-Redefine-Classes" to true,
+      "Can-Retransform-Classes" to true,
+      "Agent-Class" to agentClassName,
+      "Main-Class" to launcherClassName,
+    )
+  }
+
+  exclude("META-INF/versions/9/module-info.class")
+  duplicatesStrategy = DuplicatesStrategy.WARN
+
+  from(inline(configurations.runtimeClasspath)) // todo: remove
+}
+
+// Server running tasks
+println("----------- Agent launch config ---------------")
+println("Classpath: $serverTargetClasspath")
+println("ClassToLaunch: $serverClassToLaunch")
+println("------------------------------------------------")
+if (serverTargetClasspath != null && serverClassToLaunch != null) {
+  task("runWithAgent", JavaExec::class) {
+    group = "projector"
+    classpath(sourceSets.main.get().runtimeClasspath, tasks.jar, serverTargetClasspath)
```

Comment:
`main` property that was in `build.gradle` is missing here...

Reply:
main property is depricated

Reply:
Well, we still need to specify main class, without it JVM won't know which class to launch so the project won't work ¯\_(ツ)_/¯

So I see two options here:
1. Continue live with deprecated stuff – our project will work, but after Gradle update it can break.
2. Research how Gradle suggests changing the code, for example, by going inside the deprecated method (Ctrl+B). I see they have a way:

![image](https://user-images.githubusercontent.com/20105593/129935141-5e6ddad1-7216-4244-909e-395937da245b.png)

So it would be something like `mainClass.set` now:

![image](https://user-images.githubusercontent.com/20105593/129935275-ec9cbdfa-99c7-4a3d-9baf-f17b08c803f3.png)

Also, I believe it can be web searched too.

So let's use the following here:

```suggestion
    group = "projector"
    mainClass.set(launcherClassName)
    classpath(sourceSets.main.get().runtimeClasspath, tasks.jar, serverTargetClasspath)
```